### DISCUSSION TITLE
v0.3.0

### DISCUSSION THREAD

**Discussion by forestagostinelli** [forestagostinelli/deepxube#14] [Announcements]
* Optimizer creation and update during training now part of DeepXubeNNet class. Methods can be overridden for different behavior.
* step_min argument added to deepxube time
* Added ReLU2 activation function
* Fix start_idx usage in solve
* sample_next_state returns action taken
* Return actions taken for random_walk
* ActsRev implements method to sample reverse states, actions that return from reverse states, and transition costs along edges from resulting reverse states
* Make generating supervised labels for nodes/edges a Domain mixin
* Consolidate supervised pathfinding classes into one file.
* Sample from data buffer without replacement until all states seen
* Fix up_gen_itrs being set to 100. Now defaults to up_itrs unless specifically set to a given value.
* Separate Domain mixins/pathfinding algs for supervised node, edges, and sampling edges (for policy)
* Remove policy_rand argument
* Policy sampling method defined in PolicyNNet
* Add max_itrs arg to solve
* get_path returns list of transition costs
* Added rollout pathfinding alg, special case of beam search that does not check is_solved when searching
* PolicyNNet has _forward_train and _forward_eval abstract methods to unify forward method
* Add checkpointing during training
* Add more functionality to add neural network functions to domain and use in parallel
* Add special case of training heuristic with random policy when using pathfinding algorithm that uses a policy and not training a policy
* Add timing of getting supervised data to supervised pathfinding
* Add abstract method for loss info for training policy
* Add layer norm to resnet_fc
* Vectorize expand
* Make policy at update_num=0 sampled from Domain's sample_state_action
* Separate timings for HER and rb
* Video added to viz when visualizing solutions
* Can add random edges to graph and beam search when using policy
* Add random action option to viz
* Add --no_act option to viz
* mean(std/min/max) for itrs and path_costs in --up_v
* Make training more generic for DeepXubeNNet to allow more user flexibility
* Add gradient accumulation

<hr /><em>This discussion was created from the release <a href='https://github.com/forestagostinelli/deepxube/releases/tag/v0.3.0'>v0.3.0</a>.</em>
